alpenhorn/update.py

```diff
@@ -36,86 +39,291 @@
+                .execute()
+            )
+        except pw.DoesNotExist:
+            nodes = list()
+
+        if len(nodes) == 0:
+            log.warning(f"No active nodes on host ({host})!")
+
+        for node in nodes:
+            # Init I/O.
+            node.io.set_queue(queue)
+
+            # Check if the node is actually active
+            if not update_node_active(node):
+                continue  # Not active
+
+            # Update group info.  As we loop through the nodes, this builds up
+            # a list of which groups are available on this host and whether they
+            # were idle before node I/O happened.
+            if node.group not in groups:
+                groups[node.group] = {
+                    "nodes": [node],
+                    "idle": node.io.idle,
+                }
+            else:
+                groups[node.group]["nodes"].append(node)
+                groups[node.group]["idle"] = groups[node.group]["idle"] and node.io.idle
+
+            # Perform the node update, maybe
+            if update_node(node, queue):
+                # Remember that we updated this node
+                updated_nodes.append(node)
+
+        # Group updates
+        for group in groups:
+            if update_group(group, queue, **groups[group]):
+                # After the update, check again whether I/O is ongoing or if we're idle
+                # We need to do this here, because the node idle updates might add some
+                # stuff to the queue
+                groups[group]["idle"] = group.io.idle
+            else:
+                # If we weren't updated, then don't do the idle update either
+                groups[group]["idle"] = False
+
+        # Regular I/O updates are done.  If any nodes or groups are idle after that,
+        # run the idle updates, but only if the update happened for that group.
+
+        # loop over all the nodes which were updated, and run the idle updates if we're idle
+        for node in updated_nodes:
+            if node.io.idle:
+                node.io.idle_update()
+
+        # Ditto for groups, but we can also run the after-update hook already
+        for group in groups:
+            if groups[group]["idle"]:
+                group.io.idle_update()
+            group.io.after_update()
+
+        # loop over all the nodes again and run their after-update hooks
+        for node in nodes:
+            node.io.after_update()
+
+        # Done with the I/O updates, do some housekeeping:
 
-        # Check the time spent so far, and wait if needed
+        # Respawn workers that have exited (due to DB error)
+        pool.check()
+
+        # If we have no workers, handle some queued I/O tasks
+        if len(pool) == 0:
+            serial_io(queue)
+
+        # Check the time spent so far
         loop_time = time.time() - loop_start
-        log.info("Main loop execution was %d sec.", loop_time)
+        log.info(f"Main loop execution was {loop_time} sec.")
```

Comment:
```suggestion
        log.info(f"Main loop execution was {util.pretty_deltat(loop_time)}.")
```
(which was the whole point of introducing that function at this point.)